### ISSUE TITLE
inconsistent results when "masking" gating term between "fused_recurrent" and "fused_chunk" (fused_chunk presumably wrong)

### ISSUE THREAD

**Issue by theodorblackbird** [sustcsonglin/flash-linear-attention#20]
Hi,

I tried to implement an option where I can pass a mask to a GLA layer that  sets to zero certain hidden states so that I can pack sequences and avoid data contamination between them. For that I set corresponding ``` gk ``` to -infty so that gp.exp() = 0.
It boils down to these two lines : 

```
if reset_mask is not None:
    gk = gk.masked_fill(reset_mask, -0.1*torch.finfo(gk.dtype).max)
```
where reset_mask is a boolean mask that indicates first token of packed sequences in each row.

Apparently it works when using using "fused_recurrent" mode. In this test I compare concat(gla(x1), gla(x2)) against gla(concat(x1, x2), reset_mask=reset_mask). It also works on real data (while contaminated version fails).

```

device = "cuda"
mode = "fused_recurrent"
b, n, d = 1, 8, 64
gla = GatedLinearAttention(hidden_size=d, mode=mode).to(device)

x = torch.randn(b, n, d).to(device)
x1, x2 = x.chunk(2, dim=1)

reset_mask = (torch.arange(n, device=device)%(n//2)) == 0 #False everywhere except in first position of x1 and x2
reset_mask = rearrange(reset_mask, "n -> 1 n 1")

y = gla(x, reset_mask=reset_mask)
y1, y2 = gla(x1), gla(x2)

assert torch.allclose(torch.cat((y1, y2), dim=1), y)
```
 However "fused_chunk" curiously fails the test, and training on real data is unstable.

```
mode = "fused_recurrent"
...
>>> assert torch.allclose(torch.cat((y1, y2), dim=1), y)
AssertionError
```

Is that a bug ?

Thank you,
Théodor

**Comment by sustcsonglin** [comment]
Hello, 

I don't believe it's a bug. You might consider modifying the code as follows:
```python
if reset_mask is not None:
    gk = gk.masked_fill(reset_mask, -20)
```
and the relative error looks good to me.

Using `-0.1 * torch.finfo(gk.dtype).max` might be too large. In chunk or fused chunk mode, the kernel needs to compute the cumulative sum of log probabilities. If this value is extremely large, it could cause problems.
